polkadot/node/network/collator-protocol/src/validator_side_experimental/peer_manager/connected.rs

```diff
@@ -241,42 +273,1008 @@ impl PerPara {
+					|_, _| async { Score::default() },
+					first_peer,
+					default_connected_state()
+				)
+				.await,
+			TryAcceptOutcome::Added
+		);
+		assert_eq!(connected.peer_info(&first_peer).unwrap(), &default_connected_state());
+		for per_para in connected.per_para.values() {
+			assert!(per_para.contains(&first_peer));
+			assert_eq!(per_para.get_score(&first_peer).unwrap(), Score::default());
+		}
+
+		// Try adding an already accepted peer.
+		assert_eq!(
+			connected
+				.try_accept(
+					|_, _| async { Score::default() },
+					first_peer,
+					default_connected_state()
+				)
+				.await,
+			TryAcceptOutcome::Added
+		);
+		assert_eq!(connected.peer_info(&first_peer).unwrap(), &default_connected_state());
+		for per_para in connected.per_para.values() {
+			assert!(per_para.contains(&first_peer));
+			assert_eq!(per_para.get_score(&first_peer).unwrap(), Score::default());
+		}
+
+		// Try accepting a peer which has past reputation for an unscheduled para.
+		let second_peer = PeerId::random();
+		assert_eq!(
+			connected
+				.try_accept(
+					|peer_id, para_id| async move {
+						if peer_id == second_peer && para_id == ParaId::from(100) {
+							Score::new(10).unwrap()
+						} else {
+							Score::default()
+						}
+					},
+					second_peer,
+					default_connected_state()
+				)
+				.await,
+			TryAcceptOutcome::Added
+		);
+		assert_eq!(connected.peer_info(&first_peer).unwrap(), &default_connected_state());
+		assert_eq!(connected.peer_info(&second_peer).unwrap(), &default_connected_state());
+
+		for per_para in connected.per_para.values() {
+			assert!(per_para.contains(&second_peer));
+			assert_eq!(per_para.get_score(&second_peer).unwrap(), Score::default());
+		}
+
+		// Try accepting a peer which has past reputation for a scheduled para but is not yet
+		// declared.
+		let third_peer = PeerId::random();
+		let third_peer_para_id = ParaId::from(3);
+		assert_eq!(
+			connected
+				.try_accept(
+					|peer_id, para_id| async move {
+						if peer_id == third_peer && para_id == third_peer_para_id {
+							Score::new(10).unwrap()
+						} else {
+							Score::default()
+						}
+					},
+					third_peer,
+					default_connected_state()
+				)
+				.await,
+			TryAcceptOutcome::Added
+		);
+		assert_eq!(connected.peer_info(&first_peer).unwrap(), &default_connected_state());
+		assert_eq!(connected.peer_info(&second_peer).unwrap(), &default_connected_state());
+		assert_eq!(connected.peer_info(&third_peer).unwrap(), &default_connected_state());
+
+		for (para_id, per_para) in connected.per_para.iter() {
+			assert!(per_para.contains(&third_peer));
+
+			if para_id == &third_peer_para_id {
+				assert_eq!(per_para.get_score(&third_peer).unwrap(), Score::new(10).unwrap());
+			} else {
+				assert_eq!(per_para.get_score(&third_peer).unwrap(), Score::default());
+			}
+		}
+
+		// Try accepting a peer which is declared for an unscheduled para. It will be rejected,
+		// regardless of its other reputations.
```

Comment:
Why? As long as we have enough free spots, we should not disconnect anybody.

Reply:
They are free to connect and declare with up to `scheduling_lookahead` blocks in advance, but I don't see the benefit it adds to maintain connections for all paras if their collators declared that they are collating for a different para. The connection will be kept if they don't declare and there are enough free spots.